### PR TITLE
refactor engine

### PR DESCRIPTION
**Pull request by birdbrained**

### CODE REVIEW

**Comment by birdbrained** [comment]
Allows for the currying of functions that implement actions.

File: client/app/controllers/collection/add.js [hunk 1]
```diff
@@ -83,43 +11,125 @@ export default Ember.Controller.extend({
     }),
 
     widgets: [],
+    actions: [],
```
Comment (birdbrained) [inline client/app/controllers/collection/add.js]: as with the widgets arr, actions arr allows easier manipulation of the actions associated with the form.

File: client/app/controllers/collection/add.js [hunk 2]
```diff
@@ -83,43 +11,125 @@ export default Ember.Controller.extend({
...
+
+    // Fire enabled actions.
+    updateState: function(actions) {
```
Comment (birdbrained) [inline client/app/controllers/collection/add.js]: core logic in here. iterates over each action and runs the ones that are enabled.

File: client/app/controllers/collection/add.js [hunk 3]
```diff
@@ -83,43 +11,125 @@ export default Ember.Controller.extend({
...
+        actions.forEach((action) => {
+            // Check if the action can fire.
+            const may_fire = check_all.call(this, action.conditions);
+            if (!may_fire) return;
+            // Action may fire if execution has reached this point.
+            // Call the action and set its result and any
+            // changes to its state on `controller.parameters`.
+            console.log('stop here');
+            action.output_parameter['value'] = action.action.apply(this, action.arg_arr);
+            action.output_parameter['state'] = ['defined'];
```
Comment (birdbrained) [inline client/app/controllers/collection/add.js]: modify the obj without having to call this.set

File: client/app/controllers/collection/add.js [hunk 4]
```diff
@@ -83,43 +11,125 @@ export default Ember.Controller.extend({
...
+        });
+    },
+
+
+    // Take the description of an action and set its properties to be the vaious literal
+    // functions and parameters it depends on to operate.
+    hydrate_action: function(action) {
```
Comment (birdbrained) [inline client/app/controllers/collection/add.js]: puts everything the action needs in order to fire (literal parameters and the literal fn to be called) in one place; no calling and filtering through all sorts of controller properties.

File: client/app/controllers/collection/add.js [hunk 5]
```diff
@@ -83,43 +11,125 @@ export default Ember.Controller.extend({
...
 
-    create_widget(parameters) {
-        this.get('widgets').pushObject(parameters);
-        return parameters;
     },
 
-    run_update: function () {
-        updateState.call(this);
+    create_widget_signature: ['widget_component', 'description',
+                                        'section', 'output_parameter', 'action_id'],
```
Comment (birdbrained) [inline client/app/controllers/collection/add.js]: Signature allows fns to be curried before being passed into the component as a closure action, while remaining an obj on the model defn.

File: client/app/controllers/collection/add.js [hunk 6]
```diff
@@ -133,3 +143,100 @@ export default Ember.Controller.extend({
...
+    return !conditions.some(condition => !condition_dispatcher.call(this, condition));
+}
+
+function check_any(conditions) {
+    // If any conditions are met, the whole check passes.
+    return action.conditions.some(condition_dispatcher.bind(this))
+}
+
+
+function cons_arg_arr(action) {
```
Comment (birdbrained) [inline client/app/controllers/collection/add.js]: This is how the signature allows an obj to be used as the params for a fn. this also allows for hardcoded args as well as params to be described on the model. parameters are analagous to variables that are on the controller, args are analagous to hardcoded values that aren't dependent on the operation of the form.

File: client/app/controllers/collection/add.js [hunk 7]
```diff
@@ -133,3 +143,100 @@ export default Ember.Controller.extend({
...
+        if (typeof action.parameters === 'object' &&
+            typeof action.parameters[key] === 'string'
+        ) {
+            if (parameters[action.parameters[key]] !== 'object') {
+                console.log('no parameter, initializing with empty object');
+                parameters[action.parameters[key]] = {};
+            }
+            value = parameters[action.parameters[key]];
+        }
+        // If an arg is defined, it takes priority.
```
Comment (birdbrained) [inline client/app/controllers/collection/add.js]: if the model has an arg, it overrides the parameter that would be received from the controller.

File: client/app/initializers/promise.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import Ember from 'ember';
+
+export function initialize() {
+  window.Promise = Ember.RSVP.Promise;
```
Comment (birdbrained) [inline client/app/initializers/promise.js]: lets us use regenerator.runtime (async/await)

File: client/app/routes/collection/add.js [hunk 9]
```diff
@@ -27,7 +27,7 @@ export default Ember.Route.extend({
                 'authors',
                 'submit'
             ],
-            initial_state: {
+            initial_parameters: {
```
Comment (birdbrained) [inline client/app/routes/collection/add.js]: rename state to parameters to increase clarity

File: client/app/routes/collection/add.js [hunk 10]
```diff
@@ -148,11 +161,18 @@ export default Ember.Route.extend({
     },
 
     setupController(controller, model) {
-        controller.set('form_config', model);
-        controller.set('actions', model.actions);
+
```
Comment (birdbrained) [inline client/app/routes/collection/add.js]: improve the initialisation of the controller.

File: client/app/templates/collection/add.hbs [hunk 11]
```diff
@@ -11,7 +11,10 @@
                     <h3>{{section}}</h3>
                     {{#each widgets as |widget|}}
                         {{#if (eq widget.section section)}}
-                            {{component widget.widget_component saveParameter=(action saveParameter state widget.output)}}
+                            {{component widget.widget_component 
+                                saveParameter=(action saveParameter widget.output_parameter)
+                                action=widget.action
```
Comment (samanehsan) [inline client/app/templates/collection/add.hbs]: Is there a more descriptive name we can use here besides `action`?

File: client/app/components/button-widget/component.js [hunk 12]
```diff
@@ -7,7 +7,15 @@ export default Ember.Component.extend({
...
         pressButton() {
-            this.attrs.action();
+            try {
+                let result = this.get('action')();
+                this.attrs.saveParameter({
+                    value: result,
+                    state: ['defined']
+                });
+            } catch(ex) {
+                alert(ex);
```
Comment (samanehsan) [inline client/app/components/button-widget/component.js]: Please change this to a toast notification 🍞

File: client/app/controllers/collection/add.js (hunk 4 above)
Comment (samanehsan) [inline client/app/controllers/collection/add.js]: Please provide a more descriptive name here that indicates what the function is doing

File: client/ember-cli-build.js [hunk 13]
```diff
@@ -4,6 +4,9 @@ var EmberApp = require('ember-cli/lib/broccoli/ember-app');
 
 module.exports = function(defaults) {
   var app = new EmberApp(defaults, {
+    babel: {
+      includePolyfill: true
```
Comment (samanehsan) [inline client/ember-cli-build.js]: I got a warning when running the ember server: `
DEPRECATION: Putting the "includePolyfill" option in "babel" is deprecated, please put it in "ember-cli-babel" instead.`

File: client/app/initializers/promise.js (hunk 8 above)
Comment (samanehsan) [inline client/app/initializers/promise.js]: Why does `Ember.RSVP.Promise` need to go in an initializer if it is already accessible by importing Ember`?

File: client/app/controllers/collection/add.js [hunk 14]
```diff
@@ -1,125 +1,145 @@
...
     addMethod: 'select', // 'select' or 'create'
     methodSelected: false,
     type: Ember.computed('model.settings', function() {
         var collectionType = this.get('model.settings.collectionType') || 'project';
         return collectionType.toLowerCase();
     }),
 
+
     widgets: [],
+    actions: [],
```
Comment (samanehsan) [inline client/app/controllers/collection/add.js]: `actions` is defined in the controller here as well but there should only be one: https://github.com/cos-labs/collections/pull/81/files#diff-0fd60a923647a8bdfca3e3418a32795dR143

File: client/app/components/button-widget/component.js [hunk 15]
```diff
@@ -6,8 +6,18 @@ export default Ember.Component.extend({
     buttonString: 'Save',
 
     actions: {
-        pressButton() {
-            this.attrs.action();
+        async pressButton() {
+            try {
+                let result = await this.get('action')();
+                console.log(result);
+                debugger;
```
Comment (samanehsan) [inline client/app/components/button-widget/component.js]: Please remove the `debugger` and `console.log` statements

File: client/app/components/file-uploader/component.js [hunk 16]
```diff
@@ -1,61 +1,25 @@
...
-        return token;
-    }
-}
-
 export default Ember.Component.extend({
 
     actions: {
 
-        uploadFile: function(target) {
+        uploadFile: function(ev) {
```
Comment (samanehsan) [inline client/app/components/file-uploader/component.js]: What is `ev`? Perhaps a more descriptive name would make that clearer

File: client/app/components/file-uploader/component.js [hunk 17]
```diff
@@ -1,61 +1,25 @@
...
 
-            const saveParameter = this.attrs.saveParameter;
             const reader = new FileReader();
-            const f = event.target.files[0];
-            const uri = "https://files.osf.io/v1/resources/h8d72/providers/osfstorage/?kind=file&name=" + f.name;
```
Comment (samanehsan) [inline client/app/components/file-uploader/component.js]: Can you change this url depending on the current OSF backend? Right now file uploads will only work with production as the backend.

File: client/app/controllers/collection/add.js (hunk 6 above)
Comment (samanehsan) [inline client/app/controllers/collection/add.js]: Prefer descriptive function names over abbreviations

File: client/app/routes/collection/add.js [hunk 18]
```diff
@@ -65,15 +65,27 @@ export default Ember.Route.extend({
...
+                type: 'upload_file',
                 parameters: {
+                    file_data: 'preprint_file_data',
+                    file_name: 'preprint_file_name',
+                    node: 'preprint_node'
+                },
+                output_parameter: 'preprint_file_url',
+            }, {
+                type: 'create_widget',
+                args: {
```
Comment (samanehsan) [inline client/app/routes/collection/add.js]: What are these arguments for? Is it the action?

File: client/app/controllers/collection/add.js [hunk 19]
```diff
@@ -132,4 +151,104 @@ export default Ember.Controller.extend({
...
+}
+
+
+function check_any(conditions) {
+    // If any conditions are met, the whole check passes.
+    return conditions.some(condition_dispatcher.bind(this))
+}
+
+
+function cons_arg_arr(action) {
```
Comment (samanehsan) [inline client/app/controllers/collection/add.js]: Please provide a more descriptive function name, along with a comment about what this does including an example of what gets returned

File: client/app/initializers/promise.js (hunk 8 above)
Comment (birdbrained) [inline client/app/initializers/promise.js]: this makes the window.promise the same as the one used by ember.

File: client/app/templates/collection/add.hbs (hunk 11 above)
Comment (birdbrained) [inline client/app/templates/collection/add.hbs]: probably, but I don't know what it is.

File: client/app/components/file-uploader/component.js (hunk 16 above)
Comment (birdbrained) [inline client/app/components/file-uploader/component.js]: event

**Pull request merged**